sdk/include/opentelemetry/sdk/resource/resource.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/trace/attribute_utils.h"
+#include "opentelemetry/sdk/version/version.h"
+#include "opentelemetry/version.h"
+
+#include <cstdlib>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+using ResourceAttributes =
+    std::unordered_map<std::string, opentelemetry::sdk::trace::SpanDataAttributeValue>;
+
+class Resource
+{
+public:
+  Resource(const ResourceAttributes &attributes = ResourceAttributes()) noexcept;
```

Comment:
The specification says that *ANY* created resource should go through the environment variable additions + default logic you have in `Create`.   Should we hide this constructor to only the `OTELResourceDetector` + `getDefault` ?

Reply:
There is challenge to do this in C++. Ideally, I would like to make constructor as private, and make `OTELResourceDetector` as friend class of `Resource`, but we actually need all derivations of `ResourceDetector` access to this constructor. 
(opentelementry-java)[https://github.com/open-telemetry/opentelemetry-java/blob/f9bf909372a63bd048cfa58a384eba92005b2277/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#L30] does it using `AutoValue` (https://github.com/google/auto/tree/master/value) annotation, and (opentelemetry-dotnet)[https://github.com/open-telemetry/opentelemetry-dotnet/blob/7a91923fb7a1ee6fbc75b6fdc145137e93b38917/src/OpenTelemetry/Resources/Resource.cs#L36]  does it using `internal` keyword.

Reply:
Ok. Please ignore above comment. I have now made constructor `protected` ( so that it can be used in test cases ), and make `OTELResourceDetector` as its friend class. It seems we don't need to invoke Resource constructor in any other detector, so we should be fine.